stdlib/condition.mli

```diff
@@ -13,41 +13,164 @@
-   is done. Condition variables should always be protected by a mutex.
-   The typical use is (if [D] is a shared data structure, [m] its mutex,
-   and [c] is a condition variable):
+(**Condition variables.
+
+   Condition variables are useful when several threads wish to access a
+   shared data structure that is protected by a mutex (a mutual exclusion
+   lock).
+
+   A condition variable is a {i communication channel}. On the receiver
+   side, one or more threads can indicate that they wish to {i wait}
+   for a certain property to become true. On the sender side, a thread
+   can {i signal} that this property has become true, causing one (or
+   more) waiting threads to be woken up.
+
+   For instance, in the implementation of a queue data structure, if a
+   thread that wishes to extract an element finds that the queue is
+   currently empty, then this thread waits for the queue to become
+   nonempty. A thread that inserts an element into the queue signals
+   that the queue has become nonempty. A condition variable is used for this
+   purpose. This communication channel conveys the information that
+   the property "the queue is nonempty" is true, or more accurately,
+   may be true. (We explain below why the receiver of a signal cannot
+   be certain that the property holds.)
+
+   To continue the example of the queue, assuming that the queue has a fixed
+   maximum capacity, then a thread that wishes to insert an element
+   may find that the queue is full. Then, this thread must wait for
+   the queue to become not full, and a thread that extracts an element
+   of the queue signals that the queue has become not full. Another
+   condition variable is used for this purpose.
+
+   In short, a condition variable [c] is used to convey the information
+   that a certain property [P] about a shared data structure [D],
+   protected by a mutex [m], may be true.
+
+   Condition variables provide an efficient alternative to busy-waiting.
+   When one wishes to wait for the property [P] to be true,
+   instead of writing a busy-waiting loop:
+   {[
+     Mutex.lock m;
+     while not P do
+       Mutex.unlock m; Mutex.lock m
+     done;
+     <update the data structure>;
+     Mutex.unlock m
+   ]}
+   one uses {!wait} in the body of the loop, as follows:
    {[
      Mutex.lock m;
-     while (* some predicate P over D is not satisfied *) do
+     while not P do
        Condition.wait c m
      done;
-     (* Modify D *)
-     if (* the predicate P over D is now satisfied *) then Condition.signal c;
+     <update the data structure>;
      Mutex.unlock m
    ]}
-*)
+   The busy-waiting loop is inefficient because the waiting thread
+   consumes processing time and creates contention of the mutex [m].
+   Calling {!wait} allows the waiting thread to be suspended, so it
+   does not consume any computing resources while waiting.
+
+   With a condition variable [c], exactly one mutex [m] is associated.
```

Comment:
Yoda speak this is a bit of, no?
"Exactly one mutex  [m] is associated with a (given) condition variable [c]."
(I see there are more occurrences of the pattern below, so if you like the pattern it's fine with me too.)

Reply:
I didn't pay attention, but it sounds natural to me...

Reply:
"A condition variable has exactly one associated mutex" ?